File: 144 Binary Tree Preorder Traversal.cpp
#include <stack>
#include <vector>

Struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
};

vector<int> preorderTraversal(TreeNode* root)
{
    std::vector<int> ret;
    if(root == 0) return ret;
    
    std::stack<TreeNode*> st;
    TreeNode* p = root;
    while(!st.empty() || p != 0)
    {
        if(p != 0)
        {
            ret.push_back(p->val);
            st.push(p);
            p = p->left; 
        }
        else
        {
            p = st.top();
            st.pop();
            p = p->right;
        }
    }
    return ret;
}

File: 144 Binary Tree Preorder Traversal.ts
/*
144 Binary Tree Preorder Traversal
https://leetcode.com/problems/binary-tree-preorder-traversal
*/
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */
function preorderTraversal(root: TreeNode | null): number[] {
    const result: number[] = [];
    const stack: TreeNode[] = [];
    let curr = root;
    while (curr || stack.length > 0) {
        if (curr) {
            result.push(curr.val);
            stack.push(curr);
            curr = curr.left;
        } else {
            curr = stack.pop().right;
        }
    }
    return result;
};
